File: packages/core/src/research_engine/services/ingestion/chunking/prose_window.py
"""Prose window chunker — sentence-boundary-aware sliding window."""

from __future__ import annotations

import re

from research_engine.domain.passages import PassageDraft
from research_engine.services.ingestion.chunking.fixed_window import (
    cap_spans,
    split_at_boundary,
)
from research_engine.services.text.tokens import (
    approx_tokens,
    chars_per_token,
    token_budget_chars,
)

# Simple sentence boundary pattern
_SENT_BOUNDARY = re.compile(r"(?<=[.!?])\s+(?=[A-Z])")


def sentence_spans(text: str) -> list[tuple[int, int]]:
    """Sentence spans as ``(start, end)`` offsets into *text*.

    Boundary whitespace falls between spans rather than inside them, so slicing
    from the first sentence's start to the last sentence's end recovers the
    original inter-sentence spacing exactly.
    """
    spans: list[tuple[int, int]] = []
    start = 0
    for match in _SENT_BOUNDARY.finditer(text):
        spans.append((start, match.start()))
        start = match.end()
    if start < len(text):
        spans.append((start, len(text)))
    return spans


class ProseWindowChunker:
    id = "prose_window"
    #: What `chunk()` takes: "text" or "sections".
    consumes = "text"
    # 3.0: a unit larger than the window is broken at a word or line boundary
    # instead of emitted whole. Found in the corpus, not in a fixture: a book
    # index has almost no sentence punctuation, so the whole of one became a
    # single 22,131-token passage — well past the 8,192 the embedder accepts,
    # which means most of it was stored but never embedded.
    # 4.0: token estimates are script-aware, so the window means the same thing
    # in Greek as in English. Boundaries move only in non-Latin text; an
    # ASCII-only document chunks byte-identically to 3.0.
    version = "4.0"

    def __init__(self, max_tokens: int = 500, overlap_tokens: int = 50) -> None:
        self._max_tokens = max_tokens
        self._overlap_tokens = overlap_tokens

    @property
    def max_passage_tokens(self) -> int | None:
        return self._max_tokens

    async def chunk(self, text: str, metadata: dict | None = None) -> list[PassageDraft]:
        """Split text into overlapping chunks at sentence boundaries."""
        if not text.strip():
            return []

        # Measured once for the document: the script mix of a sentence is the
        # script mix of the book it came from, and measuring per span would put
        # an O(n) scan inside the loop below.
        rate = chars_per_token(text)

        # Sentence spans do not overlap, so capping them here is safe; the
        # overlap this chunker adds is expressed later, by widening a window's
        # start rather than by re-cutting.
        spans = cap_spans(
            text,
            [
                piece
                for span in sentence_spans(text)
                for piece in self._fit_to_window(text, span, rate)
            ],
            self._max_tokens,
        )
        if not spans:
            return []

        chunks: list[PassageDraft] = []
        window: list[tuple[int, int]] = []
        window_tokens = 0
        position = 0

        for span in spans:
            span_tokens = self._span_tokens(text, span, rate)

            if window and window_tokens + span_tokens > self._max_tokens:
                chunks.append(
                    self._make_draft(
                        text, window[0][0], window[-1][1], position, metadata, rate
                    )
                )
                position += 1
                window = self._overlap_window(text, window, rate)
                window_tokens = sum(self._span_tokens(text, s, rate) for s in window)

            window.append(span)
            window_tokens += span_tokens

        if window:
            chunks.append(
                self._make_draft(
                    text, window[0][0], window[-1][1], position, metadata, rate
                )
            )

        return chunks

    def _fit_to_window(
        self, text: str, span: tuple[int, int], rate: float
    ) -> list[tuple[int, int]]:
        """Break a unit that will not fit, at the best seam available."""
        return split_at_boundary(
            text, span[0], span[1], token_budget_chars(self._max_tokens, rate)
        )

    def _overlap_window(
        self, text: str, window: list[tuple[int, int]], rate: float
    ) -> list[tuple[int, int]]:
        """The trailing sentences of *window* that fit in the overlap budget."""
        overlap_tokens = 0
        keep_from = len(window)
        for i in range(len(window) - 1, -1, -1):
            span_tokens = self._span_tokens(text, window[i], rate)
            if overlap_tokens + span_tokens > self._overlap_tokens:
                break
            overlap_tokens += span_tokens
            keep_from = i
        return window[keep_from:]

    def _make_draft(
        self,
        text: str,
        start: int,
        end: int,
        position: int,
        metadata: dict | None,
        rate: float,
    ) -> PassageDraft:
        chunk_text = text[start:end]
        return PassageDraft(
            position=position,
            char_start=start,
            char_end=end,
            text=chunk_text,
            token_count=approx_tokens(chunk_text, rate),
            chunker=self.id,
            chunker_version=self.version,
            metadata=metadata or {},
        )

    @staticmethod
    def _span_tokens(text: str, span: tuple[int, int], rate: float) -> int:
        return approx_tokens(text[span[0] : span[1]], rate)

File: crates/marginalia-chunk/src/prose_window.rs
//! Sentence-boundary-aware sliding windows, mirroring
//! `services/ingestion/chunking/prose_window.py`.
//!
//! Chunks slice the original text from the first sentence's start to the
//! last's end, so whitespace survives and offsets stay true — the old
//! `" ".join(sentences)` rebuild collapsed every run and made spans
//! unrecoverable.

use marginalia_text::chars::is_space;
use marginalia_text::spans::{cap_spans, split_span};
use marginalia_text::tokens::{approx_tokens, chars_per_token, token_budget_chars};
use marginalia_types::sdk::PassageDraft;
use marginalia_types::Result;
use serde_json::{Map, Value};

use crate::{make_draft, CharText, ChunkerMeta};

pub const ID: &str = "prose_window";
pub const CONSUMES: &str = "text";
/// 3.0 breaks an over-window unit at a word/line seam instead of emitting it
/// whole; 4.0 measures script-aware rates, moving only non-Latin boundaries.
pub const VERSION: &str = "4.0";

/// Sentence spans as `(start, end)` char offsets into `text`.
///
/// Mirrors `_SENT_BOUNDARY = re.compile(r"(?<=[.!?])\s+(?=[A-Z])")`: after a
/// `.`/`!`/`?`, a greedy whitespace run, then an ASCII capital. The scanner is
/// equivalent by construction — the lookbehind fixes the run's start, greed
/// with backtrack reduces to "longest run, then one capital", and `finditer`
/// resumes at the run's end over characters that cannot re-trigger.
pub fn sentence_spans(text: &str) -> Vec<(usize, usize)> {
    let ct = CharText::new(text);
    sentence_spans_of(ct.chars())
}

/// [`sentence_spans`] over an already-decoded slice, so `chunk` pays the
/// decode once instead of twice.
pub(crate) fn sentence_spans_of(chars: &[char]) -> Vec<(usize, usize)> {
    let mut spans = Vec::new();
    let mut start = 0;
    let mut i = 0;
    while i < chars.len() {
        if matches!(chars[i], '.' | '!' | '?') && i + 1 < chars.len() && is_space(chars[i + 1]) {
            let mut j = i + 1;
            while j < chars.len() && is_space(chars[j]) {
                j += 1;
            }
            if j < chars.len() && chars[j].is_ascii_uppercase() {
                spans.push((start, i + 1));
                start = j;
                i = j;
                continue;
            }
        }
        i += 1;
    }
    if start < chars.len() {
        spans.push((start, chars.len()));
    }
    spans
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ProseWindowChunker {
    pub max_tokens: i64,
    pub overlap_tokens: i64,
}

impl Default for ProseWindowChunker {
    fn default() -> Self {
        Self::new(500, 50)
    }
}

impl ChunkerMeta for ProseWindowChunker {
    const ID: &'static str = ID;
    const VERSION: &'static str = VERSION;
}

impl ProseWindowChunker {
    pub fn new(max_tokens: i64, overlap_tokens: i64) -> Self {
        Self {
            max_tokens,
            overlap_tokens,
        }
    }

    pub fn max_passage_tokens(&self) -> Option<i64> {
        Some(self.max_tokens)
    }

    pub fn chunk(
        &self,
        text: &str,
        metadata: Option<&Map<String, Value>>,
    ) -> Result<Vec<PassageDraft>> {
        if text.chars().all(is_space) {
            return Ok(Vec::new());
        }
        let ct = CharText::new(text);
        // Measured once for the document: the script mix of a sentence is the
        // script mix of the book it came from; measuring per span would put an
        // O(n) scan inside the loop below.
        let rate = chars_per_token(text);
        let pieces: Vec<(usize, usize)> = sentence_spans_of(ct.chars())
            .into_iter()
            .flat_map(|span| fit_to_window(&ct, span, self.max_tokens, rate))
            .collect();
        // `cap_spans` is the one fallible pass: it rejects `max_tokens < 1`.
        let spans = cap_spans(ct.chars(), &pieces, self.max_tokens)?;

        let mut chunks = Vec::new();
        // Each span carries its count, measured once at push: the flush
        // below used to re-slice and recount every kept span, plus a fresh
        // `Vec` per flush.
        let mut window: Vec<((usize, usize), i64)> = Vec::new();
        let mut window_tokens: i64 = 0;
        let mut position: i64 = 0;
        for span in spans {
            let span_tok = tokens_of_span(&ct, span, rate);
            if !window.is_empty() && window_tokens + span_tok > self.max_tokens {
                chunks.push(make_draft::<Self>(
                    &ct,
                    window[0].0 .0,
                    window[window.len() - 1].0 .1,
                    position,
                    metadata,
                    rate,
                ));
                position += 1;
                // The trailing spans that fit the overlap budget, with their
                // memoized counts — the same scan `overlap_window` ran, minus
                // the recount.
                let mut keep_from = window.len();
                let mut kept_tokens: i64 = 0;
                for (i, &(_, t)) in window.iter().enumerate().rev() {
                    if kept_tokens + t > self.overlap_tokens {
                        break;
                    }
                    kept_tokens += t;
                    keep_from = i;
                }
                window.drain(..keep_from);
                window_tokens = kept_tokens;
            }
            window.push((span, span_tok));
            window_tokens += span_tok;
        }
        let (s, e) = (window[0].0 .0, window[window.len() - 1].0 .1);
        chunks.push(make_draft::<Self>(&ct, s, e, position, metadata, rate));
        Ok(chunks)
    }
}

/// Break a unit that will not fit, at the best seam available.
///
/// Infallible: spans come from the text being split and the budget floors at
/// 1, which is exactly [`split_span`]'s contract.
fn fit_to_window(
    ct: &CharText,
    span: (usize, usize),
    max_tokens: i64,
    rate: f64,
) -> Vec<(usize, usize)> {
    split_span(
        ct.chars(),
        span.0,
        span.1,
        token_budget_chars(max_tokens, rate).max(1) as usize,
    )
}

fn tokens_of_span(ct: &CharText, span: (usize, usize), rate: f64) -> i64 {
    approx_tokens(ct.slice(span.0, span.1), Some(rate))
}
